pkg/build/addimage/internal/container/docker/archive.go

```diff
@@ -0,0 +1,167 @@
+	tarWriter := tar.NewWriter(writer)
+	// iterate all entries in the tarball
+	for {
+		// read an entry
+		hdr, err := tarReader.Next()
+		if err == io.EOF {
+			return tarWriter.Close()
+		} else if err != nil {
+			return err
+		}
+		b, err := ioutil.ReadAll(tarReader)
+		if err != nil {
+			return err
+		}
+
+		// edit the repostories and manifests files when we find them
+		if hdr.Name == "repositories" {
+			b, err = editRepositoriesFile(b, editRepositories)
+			if err != nil {
+				return err
+			}
+			hdr.Size = int64(len(b))
+		} else if hdr.Name == "manifest.json" {
+			b, err = editManifestRepositories(b, editRepositories)
+			if err != nil {
+				return err
+			}
+			hdr.Size = int64(len(b))
+			// edit image config when we find that
+		} else if strings.HasSuffix(hdr.Name, ".json") {
+			if architectureOverride != "" {
+				b, err = editConfigArchitecture(b, architectureOverride)
+				if err != nil {
+					return err
+				}
+				hdr.Size = int64(len(b))
+			}
+		}
+
+		// write to the output tarball
+		if err := tarWriter.WriteHeader(hdr); err != nil {
+			return err
+		}
+		if len(b) > 0 {
+			if _, err := tarWriter.Write(b); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+/* helpers */
+
+func editConfigArchitecture(raw []byte, architectureOverride string) ([]byte, error) {
+	var cfg map[string]interface{}
+	if err := json.Unmarshal(raw, &cfg); err != nil {
+		return nil, err
+	}
+	const architecture = "architecture"
+	if _, ok := cfg[architecture]; !ok {
+		return raw, nil
+	}
+	cfg[architecture] = architectureOverride
+	return json.Marshal(cfg)
+}
+
+// archiveRepositories represents repository:tag:ref
+//
+// https://github.com/moby/moby/blob/master/image/spec/v1.md
+// https://github.com/moby/moby/blob/master/image/spec/v1.1.md
+// https://github.com/moby/moby/blob/master/image/spec/v1.2.md
+type archiveRepositories map[string]map[string]string
+
+func editRepositoriesFile(raw []byte, editRepositories func(string) string) ([]byte, error) {
+	tags, err := parseRepositories(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	fixed := make(archiveRepositories)
+	for repository, tagsToRefs := range tags {
+		fixed[editRepositories(repository)] = tagsToRefs
+	}
+
+	return json.Marshal(fixed)
+}
+
+// https://github.com/moby/moby/blob/master/image/spec/v1.2.md#combined-image-json--filesystem-changeset-format
+type metadataEntry struct {
+	Config   string   `json:"Config"`
+	RepoTags []string `json:"RepoTags"`
+	Layers   []string `json:"Layers"`
+}
+
+// applies
+func editManifestRepositories(raw []byte, editRepositories func(string) string) ([]byte, error) {
+	var entries []metadataEntry
+	if err := json.Unmarshal(raw, &entries); err != nil {
+		return nil, err
+	}
+
+	for i, entry := range entries {
+		fixed := make([]string, len(entry.RepoTags))
+		for i, tag := range entry.RepoTags {
+			parts := strings.Split(tag, ":")
+			if len(parts) > 2 {
+				return nil, fmt.Errorf("invalid repotag: %s", entry)
```

Comment:
This check would prevent images stored in a local registry I believe. For example:
```sh
localhost:5001/asia.gcr.io/my-gcp-project/my-docker-image:my-tag
```

Reply:
not sure if it would be an intended behavior